File: forge-app/src/jira-client.ts
/**
 * Jira API Client for Forge
 * Uses Forge's built-in API to access Jira data
 */
import api, { route, Route } from '@forge/api';

export interface JiraUser {
  accountId: string;
  displayName: string;
  emailAddress?: string;
  avatarUrls?: Record<string, string>;
}

export interface JiraAttachment {
  id: string;
  filename: string;
  size: number;
  mimeType: string;
  content: string;
  thumbnail?: string;
  created: string;
  author: JiraUser;
}

export interface JiraComment {
  id: string;
  author: JiraUser;
  body: any; // ADF format
  created: string;
  updated?: string;
}

export interface JiraIssue {
  key: string;
  id: string;
  self: string;
  fields: {
    summary: string;
    description: any;
    issuetype: { name: string; iconUrl?: string };
    status: { name: string; statusCategory?: { name: string } };
    priority: { name: string; iconUrl?: string };
    assignee?: JiraUser;
    reporter?: JiraUser;
    created: string;
    updated: string;
    resolutiondate?: string;
    fixVersions?: Array<{ name: string; released?: boolean }>;
    components?: Array<{ name: string }>;
    labels?: string[];
    attachment?: JiraAttachment[];
    comment?: { comments: JiraComment[] };
    subtasks?: Array<{ key: string; fields: { summary: string; status: { name: string } } }>;
    issuelinks?: Array<{
      type: { inward: string; outward: string };
      inwardIssue?: { key: string; fields: { summary: string } };
      outwardIssue?: { key: string; fields: { summary: string } };
    }>;
    parent?: { key: string; fields: { summary: string } };
    [key: string]: any; // Custom fields
  };
}

export interface ParsedIssue {
  key: string;
  id: string;
  summary: string;
  description: string;
  issueType: string;
  status: string;
  statusCategory?: string;
  priority: string;
  assignee?: string;
  reporter?: string;
  created: string;
  updated: string;
  resolved?: string;
  fixVersions: string[];
  components: string[];
  labels: string[];
  storyPoints?: number;
  sprints: string[];
  epic?: string;
  parent?: { key: string; summary: string };
  subtasks: Array<{ key: string; summary: string; status: string }>;
  links: Array<{ type: string; key: string; summary: string }>;
  attachments: JiraAttachment[];
  comments: Array<{ author: string; body: string; created: string }>;
  customFields: Record<string, any>;
}

/**
 * Hämta en issue med alla fält
 */
export async function getIssue(issueKey: string): Promise<ParsedIssue> {
  const response = await api.asApp().requestJira(
    route`/rest/api/3/issue/${issueKey}?expand=renderedFields,names&fields=*all`
  );
  
  if (!response.ok) {
    throw new Error(`Failed to fetch issue: ${response.status}`);
  }
  
  const data: JiraIssue = await response.json();
  
  // Hämta fältnamn för bättre visning
  let fieldNames: Record<string, string> = {};
  try {
    fieldNames = await getFieldNames();
  } catch (e) {
    console.log('Could not fetch field names, using raw keys');
  }
  
  return parseIssueWithFieldNames(data, fieldNames);
}

/**
 * Sök efter issues med JQL (POST /rest/api/3/search/jql)
 */
export async function searchIssues(jql: string, maxResults: number = 50): Promise<ParsedIssue[]> {
  const issues: ParsedIssue[] = [];
  let startAt = 0;
  
  while (issues.length < maxResults) {
    const batchSize = Math.min(50, maxResults - issues.length);
    
    // Bygg request body enligt Atlassian migration guide
    const requestBody = {
      jql,
      maxResults: batchSize,
      startAt,
      fields: [
        'summary',
        'description', 
        'issuetype',
        'status',
        'priority',
        'assignee',
        'reporter',
        'created',
        'updated',
        'resolutiondate',
        'fixVersions',
        'components',
        'labels',
        'attachment',
        'comment',
        'subtasks',
        'issuelinks',
        'parent'
      ]
    };
    
    console.log('Sending search request:', JSON.stringify(requestBody, null, 2));
    
    // Använd asUser() för att köra med användarens behörigheter
    const response = await api.asUser().requestJira(route`/rest/api/3/search/jql`, {
      method: 'POST',
      headers: {
        'Accept': 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(requestBody)
    });
    
    console.log('Response status:', response.status);
    
    if (!response.ok) {
      const errorText = await response.text();
      console.error('Search API error:', response.status, errorText);
      throw new Error(`Failed to search issues: ${response.status} - ${errorText}`);
    }
    
    const data = await response.json();
    console.log('Found issues:', data.issues?.length || 0, 'of', data.total);
    
    for (const issue of data.issues || []) {
      issues.push(parseIssue(issue));
    }
    
    if (!data.issues || data.issues.length === 0 || issues.length >= data.total) {
      break;
    }
    
    startAt += data.issues.length;
  }
  
  return issues;
}

/**
 * Ladda ner en bilaga via attachment ID
 * Använder /rest/api/3/attachment/content/{id} med rätt redirect-hantering
 */
export async function downloadAttachment(attachmentId: string): Promise<ArrayBuffer> {
  console.log('Downloading attachment ID:', attachmentId);
  
  // Först hämta attachment metadata för att få content URL
  const metaResponse = await api.asApp().requestJira(route`/rest/api/3/attachment/${attachmentId}`);
  
  if (!metaResponse.ok) {
    console.error('Failed to get attachment metadata:', metaResponse.status);
    throw new Error(`Failed to get attachment metadata: ${metaResponse.status}`);
  }
  
  const metadata = await metaResponse.json();
  console.log('Attachment metadata:', metadata.filename, 'mimeType:', metadata.mimeType);
  
  // Använd content URL direkt med media API
  // Forge kräver att vi använder requestJira för alla Jira-anrop
  const contentResponse = await api.asApp().requestJira(route`/rest/api/3/attachment/content/${attachmentId}`, {
    headers: { 
      'Accept': metadata.mimeType || 'image/png'
    }
  });
  
  console.log('Content response status:', contentResponse.status);
  
  if (!contentResponse.ok) {
    // Försök med thumbnail istället
    console.log('Trying thumbnail...');
    const thumbResponse = await api.asApp().requestJira(route`/rest/api/3/attachment/thumbnail/${attachmentId}`, {
      headers: { 
        'Accept': 'image/png'
      }
    });
    
    console.log('Thumbnail response status:', thumbResponse.status);
    
    if (!thumbResponse.ok) {
      throw new Error(`Failed to download attachment: ${contentResponse.status}`);
    }
    
    return thumbResponse.arrayBuffer();
  }
  
  return contentResponse.arrayBuffer();
}

/**
 * Hämta fältnamn för custom fields
 */
export async function getFieldNames(): Promise<Record<string, string>> {
  const response = await api.asApp().requestJira(route`/rest/api/3/field`);
  
  if (!response.ok) {
    throw new Error(`Failed to fetch fields: ${response.status}`);
  }
  
  const fields = await response.json();
  const fieldMap: Record<string, string> = {};
  
  for (const field of fields) {
    fieldMap[field.id] = field.name;
  }
  
  return fieldMap;
}

/**
 * Parsa en Jira issue till strukturerad data
 */
function parseIssue(data: JiraIssue): ParsedIssue {
  return parseIssueWithFieldNames(data, {});
}

/**
 * Parsa en Jira issue med fältnamn
 */
function parseIssueWithFieldNames(data: JiraIssue, fieldNames: Record<string, string>): ParsedIssue {
  const fields = data.fields;
  
  // Extrahera custom fields med läsbara namn
  const customFields: Record<string, any> = {};
  const skipPatterns = ['sprint', 'story point', 'epic', 'rank', 'flagged', 'development', 'team', 'color', 'change'];
  
  for (const [key, value] of Object.entries(fields)) {
    if (!key.startsWith('customfield_') || value === null || value === undefined) continue;
    
    // Hämta läsbart fältnamn
    const fieldName = fieldNames[key] || key;
    const fieldNameLower = fieldName.toLowerCase();
    
    // Hoppa över system-/irrelevanta fält
    if (skipPatterns.some(p => fieldNameLower.includes(p))) continue;
    
    // Formatera värdet
    let formattedValue: any = null;
    
    if (typeof value === 'string') {
      if (value.trim()) formattedValue = value;
    } else if (typeof value === 'number') {
      formattedValue = String(value);
    } else if (typeof value === 'object') {
      if (Array.isArray(value)) {
        const items = value.map(v => {
          if (typeof v === 'string') return v;
          if (typeof v === 'object' && v !== null) {
            return v.value || v.name || v.displayName || '';
          }
          return String(v);
        }).filter(s => s);
        if (items.length) formattedValue = items.join(', ');
      } else if (value.type === 'doc') {
        formattedValue = extractText(value);
      } else if ('value' in value) {
        formattedValue = value.value;
      } else if ('name' in value) {
        formattedValue = value.name;
      } else if ('displayName' in value) {
        formattedValue = value.displayName;
      } else if ('content' in value) {
        formattedValue = extractText(value);
      }
    }
    
    if (formattedValue !== null && formattedValue !== '' && formattedValue !== undefined) {
      customFields[fieldName] = formattedValue;
    }
  }
  
  return {
    key: data.key,
    id: data.id,
    summary: fields.summary || '',
    description: extractText(fields.description),
    issueType: fields.issuetype?.name || 'Unknown',
    status: fields.status?.name || 'Unknown',
    statusCategory: fields.status?.statusCategory?.name,
    priority: fields.priority?.name || 'None',
    assignee: fields.assignee?.displayName,
    reporter: fields.reporter?.displayName,
    created: fields.created,
    updated: fields.updated,
    resolved: fields.resolutiondate,
    fixVersions: (fields.fixVersions || []).map(v => v.name),
    components: (fields.components || []).map(c => c.name),
    labels: fields.labels || [],
    storyPoints: findCustomField(fields, 'story points', 'story point'),
    sprints: extractSprints(fields),
    epic: findCustomField(fields, 'epic'),
    parent: fields.parent ? {
      key: fields.parent.key,
      summary: fields.parent.fields?.summary || ''
    } : undefined,
    subtasks: (fields.subtasks || []).map(st => ({
      key: st.key,
      summary: st.fields?.summary || '',
      status: st.fields?.status?.name || 'Unknown'
    })),
    links: parseIssueLinks(fields.issuelinks || []),
    attachments: fields.attachment || [],
    comments: (fields.comment?.comments || []).map(c => ({
      author: c.author?.displayName || 'Unknown',
      body: extractText(c.body),
      created: c.created
    })),
    customFields
  };
}

/**
 * Extrahera text från ADF (Atlassian Document Format)
 */
function extractText(content: any): string {
  if (!content) return '';
  if (typeof content === 'string') return content;
  if (content.type !== 'doc') return String(content);
  
  return parseAdfContent(content.content || []);
}

function parseAdfContent(nodes: any[]): string {
  const result: string[] = [];
  
  for (const node of nodes) {
    switch (node.type) {
      case 'text':
        result.push(node.text || '');
        break;
      case 'paragraph':
        result.push(parseAdfContent(node.content || []) + '\n');
        break;
      case 'heading':
        const level = node.attrs?.level || 1;
        result.push('\n' + '#'.repeat(level) + ' ' + parseAdfContent(node.content || []) + '\n');
        break;
      case 'bulletList':
        for (const item of node.content || []) {
          result.push('• ' + parseAdfContent(item.content || []).trim() + '\n');
        }
        break;
      case 'orderedList':
        let i = 1;
        for (const item of node.content || []) {
          result.push(`${i++}. ` + parseAdfContent(item.content || []).trim() + '\n');
        }
        break;
      case 'listItem':
        result.push(parseAdfContent(node.content || []));
        break;
      case 'codeBlock':
        result.push('\n```\n' + parseAdfContent(node.content || []) + '\n```\n');
        break;
      case 'hardBreak':
        result.push('\n');
        break;
      case 'mention':
        result.push('@' + (node.attrs?.text || 'user'));
        break;
      default:
        if (node.content) {
          result.push(parseAdfContent(node.content));
        }
    }
  }
  
  return result.join('');
}

/**
 * Hitta custom field baserat på namn
 */
function findCustomField(fields: Record<string, any>, ...searchTerms: string[]): any {
  for (const [key, value] of Object.entries(fields)) {
    if (!key.startsWith('customfield_') || value === null) continue;
    
    const valueStr = JSON.stringify(value).toLowerCase();
    for (const term of searchTerms) {
      if (valueStr.includes(term.toLowerCase())) {
        if (typeof value === 'object' && value !== null) {
          return value.value || value.name || value;
        }
        return value;
      }
    }
  }
  return undefined;
}

/**
 * Extrahera sprint-information
 */
function extractSprints(fields: Record<string, any>): string[] {
  const sprints: string[] = [];
  
  for (const [key, value] of Object.entries(fields)) {
    if (!key.startsWith('customfield_') || !Array.isArray(value)) continue;
    
    for (const item of value) {
      if (item && typeof item === 'object' && 'name' in item) {
        if (item.name?.toLowerCase().includes('sprint') || key.toLowerCase().includes('sprint')) {
          sprints.push(item.name);
        }
      }
    }
  }
  
  return sprints;
}

/**
 * Parsa issue-länkar
 */
function parseIssueLinks(links: any[]): Array<{ type: string; key: string; summary: string }> {
  const parsed: Array<{ type: string; key: string; summary: string }> = [];
  
  for (const link of links) {
    if (link.outwardIssue) {
      parsed.push({
        type: link.type?.outward || 'relates to',
        key: link.outwardIssue.key,
        summary: link.outwardIssue.fields?.summary || ''
      });
    }
    if (link.inwardIssue) {
      parsed.push({
        type: link.type?.inward || 'relates to',
        key: link.inwardIssue.key,
        summary: link.inwardIssue.fields?.summary || ''
      });
    }
  }
  
  return parsed;
}

/**
 * Extrahera alla custom fields med deras namn
 */
async function extractCustomFieldsWithNames(fields: Record<string, any>): Promise<Record<string, any>> {
  const customFields: Record<string, any> = {};
  const skipPatterns = ['sprint', 'story point', 'epic', 'rank', 'flagged', 'development', 'team'];
  
  // Hämta fältnamn
  let fieldNames: Record<string, string> = {};
  try {
    fieldNames = await getFieldNames();
  } catch (e) {
    console.log('Could not fetch field names');
  }
  
  for (const [key, value] of Object.entries(fields)) {
    if (!key.startsWith('customfield_') || value === null || value === undefined) continue;
    
    // Hämta fältnamn
    const fieldName = fieldNames[key] || key;
    const fieldNameLower = fieldName.toLowerCase();
    
    // Hoppa över redan hanterade fält
    if (skipPatterns.some(p => fieldNameLower.includes(p))) continue;
    
    // Formatera värdet
    let formattedValue: any = null;
    
    if (typeof value === 'string') {
      formattedValue = value;
    } else if (typeof value === 'number') {
      formattedValue = value;
    } else if (typeof value === 'object') {
      if (Array.isArray(value)) {
        formattedValue = value.map(v => 
          typeof v === 'object' ? (v.value || v.name || v.displayName || JSON.stringify(v)) : String(v)
        ).join(', ');
      } else if ('value' in value) {
        formattedValue = value.value;
      } else if ('name' in value) {
        formattedValue = value.name;
      } else if ('displayName' in value) {
        formattedValue = value.displayName;
      } else if ('content' in value) {
        formattedValue = extractText(value);
      } else if (value.type === 'doc') {
        formattedValue = extractText(value);
      }
    }
    
    if (formattedValue !== null && formattedValue !== '' && formattedValue !== undefined) {
      customFields[fieldName] = formattedValue;
    }
  }
  
  return customFields;
}

/**
 * Extrahera alla custom fields (sync version)
 */
function extractCustomFields(fields: Record<string, any>): Record<string, any> {
  const customFields: Record<string, any> = {};
  const skipPatterns = ['sprint', 'story point', 'epic', 'rank', 'flagged', 'development', 'team'];
  
  for (const [key, value] of Object.entries(fields)) {
    if (!key.startsWith('customfield_') || value === null || value === undefined) continue;
    
    // Hoppa över redan hanterade fält baserat på innehåll
    const valueStr = JSON.stringify(value).toLowerCase();
    if (skipPatterns.some(p => valueStr.includes(p))) continue;
    
    // Formatera värdet
    let formattedValue: any = null;
    
    if (typeof value === 'string') {
      formattedValue = value;
    } else if (typeof value === 'number') {
      formattedValue = value;
    } else if (typeof value === 'object') {
      if (Array.isArray(value)) {
        formattedValue = value.map(v => 
          typeof v === 'object' ? (v.value || v.name || v.displayName || JSON.stringify(v)) : String(v)
        ).join(', ');
      } else if ('value' in value) {
        formattedValue = value.value;
      } else if ('name' in value) {
        formattedValue = value.name;
      } else if ('displayName' in value) {
        formattedValue = value.displayName;
      } else if ('content' in value) {
        formattedValue = extractText(value);
      } else if (value.type === 'doc') {
        formattedValue = extractText(value);
      }
    }
    
    if (formattedValue !== null && formattedValue !== '' && formattedValue !== undefined) {
      customFields[key] = formattedValue;
    }
  }
  
  return customFields;
}

File: server/jira_client.py
"""
Jira API-klient för att hämta tickets och bilagor
Uppdaterad för Jira API v3
"""
import os
import requests
from typing import Dict, List, Any, Optional
from config import Config


class JiraClient:
    """Klient för Jira API v3-interaktion"""
    
    def __init__(self, jira_url: str = None, email: str = None, api_token: str = None):
        """
        Initiera Jira-klienten
        
        Args:
            jira_url: Jira URL (default: från Config)
            email: Jira email (default: från Config)
            api_token: Jira API token (default: från Config)
        """
        self.base_url = (jira_url or Config.JIRA_URL).rstrip('/')
        self.session = requests.Session()
        self.session.auth = (email or Config.JIRA_EMAIL, api_token or Config.JIRA_API_TOKEN)
        self.session.headers.update({
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        })
        
        # Cache för fältnamn
        self._field_names = None
    
    def _get(self, endpoint: str, params: dict = None) -> dict:
        """Gör GET-request till Jira API"""
        url = f"{self.base_url}/rest/api/3/{endpoint}"
        response = self.session.get(url, params=params)
        response.raise_for_status()
        return response.json()
    
    def _get_field_names(self) -> Dict[str, str]:
        """Hämta mappning av fält-ID till namn"""
        if self._field_names is None:
            fields = self._get('field')
            self._field_names = {f['id']: f['name'] for f in fields}
        return self._field_names
    
    def get_issue(self, issue_key: str) -> Dict[str, Any]:
        """
        Hämta en specifik issue med alla fält
        
        Args:
            issue_key: Jira issue-nyckel (t.ex. SOMU-31)
            
        Returns:
            Dictionary med all issue-data
        """
        data = self._get(f'issue/{issue_key}', params={
            'expand': 'renderedFields,names,changelog',
            'fields': '*all'
        })
        return self._parse_issue(data)
    
    def get_issues_by_jql(self, jql: str, max_results: int = 100) -> List[Dict[str, Any]]:
        """
        Hämta issues baserat på JQL-query
        
        Args:
            jql: JQL-sökfråga
            max_results: Max antal resultat
            
        Returns:
            Lista med issue-data
        """
        issues = []
        next_page_token = None
        
        while len(issues) < max_results:
            params = {
                'jql': jql,
                'maxResults': min(50, max_results - len(issues)),
                'fields': '*all'
            }
            if next_page_token:
                params['nextPageToken'] = next_page_token
            
            data = self._get('search/jql', params=params)
            
            # Hämta fullständig data för varje issue
            for issue_stub in data.get('issues', []):
                issue_id = issue_stub.get('id')
                if issue_id:
                    try:
                        full_issue = self._get(f'issue/{issue_id}', params={
                            'expand': 'renderedFields,names,changelog',
                            'fields': '*all'
                        })
                        issues.append(self._parse_issue(full_issue))
                    except Exception as e:
                        print(f"⚠️  Kunde inte hämta issue {issue_id}: {e}")
            
            # Kolla om det finns fler sidor
            next_page_token = data.get('nextPageToken')
            if data.get('isLast', True) or not next_page_token:
                break
        
        return issues
    
    def get_project_issues(self, project_key: str, issue_type: str = None) -> List[Dict[str, Any]]:
        """
        Hämta alla issues för ett projekt
        
        Args:
            project_key: Projektets nyckel
            issue_type: Filtrera på issue-typ (t.ex. "Story", "Bug")
            
        Returns:
            Lista med issue-data
        """
        jql = f'project = {project_key}'
        if issue_type:
            jql += f' AND issuetype = "{issue_type}"'
        jql += ' ORDER BY key DESC'
        
        return self.get_issues_by_jql(jql)
    
    def _parse_issue(self, data: dict) -> Dict[str, Any]:
        """
        Parsa en Jira-issue till en strukturerad dictionary
        
        Args:
            data: Rå issue-data från API
            
        Returns:
            Strukturerad dictionary med issue-data
        """
        fields = data.get('fields', {})
        
        # Grundläggande fält
        issue_data = {
            'key': data.get('key'),
            'id': data.get('id'),
            'self': data.get('self'),
            'summary': fields.get('summary', ''),
            'description': self._extract_text(fields.get('description')),
            'rendered_description': '',
            'issue_type': {
                'name': fields.get('issuetype', {}).get('name', 'Unknown') if fields.get('issuetype') else 'Unknown',
                'icon_url': fields.get('issuetype', {}).get('iconUrl') if fields.get('issuetype') else None
            },
            'status': {
                'name': fields.get('status', {}).get('name', 'Unknown') if fields.get('status') else 'Unknown',
                'category': fields.get('status', {}).get('statusCategory', {}).get('name') if fields.get('status') else None
            },
            'priority': {
                'name': fields.get('priority', {}).get('name', 'None') if fields.get('priority') else 'None',
                'icon_url': fields.get('priority', {}).get('iconUrl') if fields.get('priority') else None
            },
            'created': fields.get('created'),
            'updated': fields.get('updated'),
            'resolved': fields.get('resolutiondate'),
        }
        
        # Tilldelning och rapportör
        issue_data['assignee'] = self._parse_user(fields.get('assignee'))
        issue_data['reporter'] = self._parse_user(fields.get('reporter'))
        
        # Story Points (vanliga custom field-namn)
        issue_data['story_points'] = self._find_custom_field(fields, 
            ['story points', 'storypoints', 'story point estimate'])
        
        # Fix versions
        issue_data['fix_versions'] = [
            {'name': v.get('name', ''), 'released': v.get('released', False)}
            for v in (fields.get('fixVersions') or [])
        ]
        
        # Components
        issue_data['components'] = [c.get('name', '') for c in (fields.get('components') or [])]
        
        # Labels
        issue_data['labels'] = fields.get('labels') or []
        
        # Sprint information
        issue_data['sprints'] = self._extract_sprints(fields)
        
        # Epic link
        issue_data['epic'] = self._find_custom_field(fields, ['epic link', 'epic name', 'parent epic'])
        
        # Parent (för subtasks)
        parent = fields.get('parent')
        issue_data['parent'] = {
            'key': parent.get('key'),
            'summary': parent.get('fields', {}).get('summary', '')
        } if parent else None
        
        # Subtasks
        issue_data['subtasks'] = [
            {
                'key': st.get('key'),
                'summary': st.get('fields', {}).get('summary', ''),
                'status': st.get('fields', {}).get('status', {}).get('name', 'Unknown')
            }
            for st in (fields.get('subtasks') or [])
        ]
        
        # Länkade issues
        issue_data['links'] = self._parse_issue_links(fields.get('issuelinks') or [])
        
        # Bilagor
        issue_data['attachments'] = [
            {
                'id': att.get('id'),
                'filename': att.get('filename'),
                'size': att.get('size', 0),
                'mime_type': att.get('mimeType', ''),
                'content_url': att.get('content'),
                'thumbnail_url': att.get('thumbnail'),
                'created': att.get('created'),
                'author': att.get('author', {}).get('displayName', 'Unknown') if att.get('author') else 'Unknown'
            }
            for att in (fields.get('attachment') or [])
        ]
        
        # Kommentarer
        comment_data = fields.get('comment', {})
        comments = comment_data.get('comments', []) if isinstance(comment_data, dict) else []
        issue_data['comments'] = [
            {
                'id': comment.get('id'),
                'author': comment.get('author', {}).get('displayName', 'Unknown') if comment.get('author') else 'Unknown',
                'body': self._extract_text(comment.get('body')),
                'created': comment.get('created'),
                'updated': comment.get('updated')
            }
            for comment in comments
        ]
        
        # Alla custom fields
        issue_data['custom_fields'] = self._get_all_custom_fields(fields)
        
        return issue_data
    
    def _extract_text(self, content) -> str:
        """Extrahera text från Atlassian Document Format (ADF)"""
        if content is None:
            return ''
        if isinstance(content, str):
            return content
        if not isinstance(content, dict):
            return str(content)
        
        # ADF-format
        if content.get('type') == 'doc':
            return self._parse_adf_content(content.get('content', []))
        
        return str(content)
    
    def _parse_adf_content(self, content_list: list) -> str:
        """Parsa ADF content till text"""
        result = []
        
        for node in content_list:
            node_type = node.get('type', '')
            
            if node_type == 'text':
                result.append(node.get('text', ''))
            
            elif node_type == 'paragraph':
                para_text = self._parse_adf_content(node.get('content', []))
                result.append(para_text + '\n')
            
            elif node_type == 'heading':
                heading_text = self._parse_adf_content(node.get('content', []))
                level = node.get('attrs', {}).get('level', 1)
                result.append(f"\n{'#' * level} {heading_text}\n")
            
            elif node_type == 'bulletList':
                for item in node.get('content', []):
                    item_text = self._parse_adf_content(item.get('content', []))
                    result.append(f"* {item_text.strip()}\n")
            
            elif node_type == 'orderedList':
                for i, item in enumerate(node.get('content', []), 1):
                    item_text = self._parse_adf_content(item.get('content', []))
                    result.append(f"{i}. {item_text.strip()}\n")
            
            elif node_type == 'listItem':
                result.append(self._parse_adf_content(node.get('content', [])))
            
            elif node_type == 'codeBlock':
                code_text = self._parse_adf_content(node.get('content', []))
                result.append(f"\n```\n{code_text}\n```\n")
            
            elif node_type == 'blockquote':
                quote_text = self._parse_adf_content(node.get('content', []))
                result.append(f"> {quote_text}")
            
            elif node_type == 'hardBreak':
                result.append('\n')
            
            elif node_type == 'mention':
                result.append(f"@{node.get('attrs', {}).get('text', 'user')}")
            
            elif node_type == 'emoji':
                result.append(node.get('attrs', {}).get('text', ''))
            
            elif node_type == 'inlineCard' or node_type == 'blockCard':
                url = node.get('attrs', {}).get('url', '')
                result.append(f"[{url}]")
            
            elif node_type == 'mediaGroup' or node_type == 'mediaSingle':
                result.append('[Media]')
            
            elif 'content' in node:
                result.append(self._parse_adf_content(node.get('content', [])))
        
        return ''.join(result)
    
    def _parse_user(self, user: dict) -> Optional[Dict[str, str]]:
        """Parsa användarinformation"""
        if not user:
            return None
        return {
            'name': user.get('displayName', 'Unknown'),
            'email': user.get('emailAddress'),
            'avatar_url': user.get('avatarUrls', {}).get('48x48') if user.get('avatarUrls') else None
        }
    
    def _find_custom_field(self, fields: dict, search_terms: list) -> Any:
        """Sök efter custom field baserat på namn"""
        field_names = self._get_field_names()
        
        for field_id, field_name in field_names.items():
            if field_id in fields and fields[field_id] is not None:
                field_name_lower = field_name.lower()
                for term in search_terms:
                    if term.lower() in field_name_lower:
                        value = fields[field_id]
                        # Hantera olika typer av värden
                        if isinstance(value, dict):
                            return value.get('value') or value.get('name') or value
                        return value
        return None
    
    def _extract_sprints(self, fields: dict) -> List[Dict[str, Any]]:
        """Extrahera sprint-information"""
        sprints = []
        field_names = self._get_field_names()
        
        for field_id, field_name in field_names.items():
            if 'sprint' in field_name.lower() and field_id in fields:
                value = fields[field_id]
                if isinstance(value, list):
                    for item in value:
                        if isinstance(item, dict):
                            sprints.append({
                                'name': item.get('name', 'Unknown'),
                                'state': item.get('state', 'unknown')
                            })
                        elif isinstance(item, str):
                            sprints.append({'name': item, 'state': 'unknown'})
        return sprints
    
    def _parse_issue_links(self, links: list) -> List[Dict[str, Any]]:
        """Parsa issue-länkar"""
        parsed_links = []
        for link in links:
            link_type = link.get('type', {})
            
            if link.get('outwardIssue'):
                outward = link['outwardIssue']
                parsed_links.append({
                    'type': link_type.get('outward', 'relates to'),
                    'key': outward.get('key'),
                    'summary': outward.get('fields', {}).get('summary', '')
                })
            
            if link.get('inwardIssue'):
                inward = link['inwardIssue']
                parsed_links.append({
                    'type': link_type.get('inward', 'relates to'),
                    'key': inward.get('key'),
                    'summary': inward.get('fields', {}).get('summary', '')
                })
        
        return parsed_links
    
    def _get_all_custom_fields(self, fields: dict) -> Dict[str, Any]:
        """Hämta alla custom fields med deras namn"""
        custom_fields = {}
        field_names = self._get_field_names()
        
        # Fält att hoppa över (redan hanterade)
        skip_patterns = ['story point', 'sprint', 'epic', 'rank', 'flagged', 
                        'development', 'team', 'change reason']
        
        for field_id, value in fields.items():
            if not field_id.startswith('customfield_'):
                continue
            if value is None:
                continue
            
            field_name = field_names.get(field_id, field_id)
            
            # Hoppa över redan hanterade fält
            if any(pattern in field_name.lower() for pattern in skip_patterns):
                continue
            
            # Hantera olika typer av värden
            if isinstance(value, dict):
                if 'value' in value:
                    custom_fields[field_name] = value['value']
                elif 'name' in value:
                    custom_fields[field_name] = value['name']
                elif 'displayName' in value:
                    custom_fields[field_name] = value['displayName']
                elif 'content' in value:  # ADF
                    custom_fields[field_name] = self._extract_text(value)
                else:
                    custom_fields[field_name] = str(value)
            elif isinstance(value, list):
                if value:
                    items = []
                    for v in value:
                        if isinstance(v, dict):
                            items.append(v.get('value') or v.get('name') or str(v))
                        else:
                            items.append(str(v))
                    custom_fields[field_name] = items
            else:
                custom_fields[field_name] = value
        
        return custom_fields
    
    def download_attachment(self, attachment: Dict[str, Any], output_dir: str) -> Optional[str]:
        """
        Ladda ner en bilaga
        
        Args:
            attachment: Attachment-dictionary från parsed issue
            output_dir: Mapp att spara filen i
            
        Returns:
            Sökväg till nedladdad fil eller None vid fel
        """
        os.makedirs(output_dir, exist_ok=True)
        
        try:
            response = self.session.get(attachment['content_url'])
            response.raise_for_status()
            
            filepath = os.path.join(output_dir, attachment['filename'])
            with open(filepath, 'wb') as f:
                f.write(response.content)
            
            return filepath
        except Exception as e:
            print(f"⚠️  Kunde inte ladda ner {attachment['filename']}: {e}")
            return None
    
    def download_all_attachments(self, issue_data: Dict[str, Any], output_dir: str) -> List[str]:
        """
        Ladda ner alla bilagor för en issue
        
        Args:
            issue_data: Parsed issue-data
            output_dir: Mapp att spara filerna i
            
        Returns:
            Lista med sökvägar till nedladdade filer
        """
        downloaded = []
        for attachment in issue_data.get('attachments', []):
            filepath = self.download_attachment(attachment, output_dir)
            if filepath:
                downloaded.append(filepath)
        return downloaded
